code/_onclick/hud/fullscreen.dm

```diff
@@ -1,125 +1,104 @@
+#define SHOULD_SHOW_TO(mymob, myscreen) (!(mymob.stat == DEAD && !myscreen.show_when_dead))
 
 /mob
 	var/list/screens = list()
 
 /mob/proc/set_fullscreen(condition, screen_name, screen_type, arg)
 	condition ? overlay_fullscreen(screen_name, screen_type, arg) : clear_fullscreen(screen_name)
 
-/mob/proc/overlay_fullscreen(category, type, severity)
+/mob/proc/overlay_fullscreen_timer(duration, animated, category, type, severity)
+	overlay_fullscreen(category, type, severity)
+	addtimer(CALLBACK(src, .proc/clear_fullscreen, category, animated), duration)
 
-	var/rescale = FALSE
+/mob/proc/overlay_fullscreen(category, type, severity)
 	var/obj/screen/fullscreen/screen = screens[category]
-
-	if(!client) //The mob needs to be player-controlled to modify the screen
-
-		return
-
-	if (client.temp_view != world.view)
-		rescale = TRUE
-
-
-
-	if(screen)
-		if(screen.type != type)
-			clear_fullscreen(category, FALSE)
-			screen = null
-		else if(!severity || severity == screen.severity)
-			return null
-
-
-	if(!screen)
-		screen = new type()
-		screen.owner = src
-
+	if (!screen || screen.type != type)
+		// needs to be recreated
+		clear_fullscreen(category, FALSE)
+		screens[category] = screen = new type()
+	else if ((!severity || severity == screen.severity) && (!client || screen.fs_view == client.view))
+		// doesn't need to be updated
+		return screen
 
 	screen.icon_state = "[initial(screen.icon_state)][severity]"
 	screen.severity = severity
-
-	if (rescale)
-		screen.set_size(client)
-
-	screens[category] = screen
-	if(client && (stat != DEAD || screen.allstate))
-
+	if (client && SHOULD_SHOW_TO(src, screen))
+		screen.update_for_view(client.view)
 		client.screen += screen
+
 	return screen
 
 /mob/proc/clear_fullscreen(category, animated = 10)
-
 	var/obj/screen/fullscreen/screen = screens[category]
-
 	if(!screen)
 		return
 
 	screens -= category
 
 	if(animated)
-		spawn(0)
-			animate(screen, alpha = 0, time = animated)
-			sleep(animated)
-			if(client)
-				client.screen -= screen
-			qdel(screen)
+		animate(screen, alpha = 0, time = animated)
+		addtimer(CALLBACK(src, .proc/clear_fullscreen_after_animate, screen), animated, TIMER_CLIENT_TIME)
 	else
 		if(client)
 			client.screen -= screen
 		qdel(screen)
 
+/mob/proc/clear_fullscreen_after_animate(obj/screen/fullscreen/screen)
+	if(client)
+		client.screen -= screen
+	qdel(screen)
+
 /mob/proc/clear_fullscreens()
 	for(var/category in screens)
-		clear_fullscreen(category, 0)
+		clear_fullscreen(category)
 
 /mob/proc/hide_fullscreens()
 	if(client)
 		for(var/category in screens)
 			client.screen -= screens[category]
 
-/mob/proc/reload_fullscreen()
+/mob/proc/reload_fullscreens()
 	if(client)
+		var/obj/screen/fullscreen/screen
```

Comment:
why are you doing this here? it should be inside the loop

Reply:
> why are you doing this here? it should be inside the loop


It is not necessary to contain it in a loop. And it seems like it does not reassign a memory cell every time. Not noticeable, but a little nice.